bug: API conversation history file truncated to empty array during task resume

When resuming a task while opening and closing orchestrator tasks, the API conversation history file is sometimes truncated to an empty array `[]`.

I am speculating that there was a race condition which caused multiple simultaneous writes to  `api_conversation_history.json` when I was clicking between tasks, and https://github.com/RooCodeInc/Roo-Code/pull/3772 will probably fix this because it is atomic and has a lock to prevent simultaneous activity on the same file.

## Error
```
mainThreadExtensionService.ts:78 [RooVeterinaryInc.roo-cline]Unexpected: No existing API conversation history
mainThreadExtensionService.ts:79 Error: Unexpected: No existing API conversation history
    at _Task.resumeTaskFromHistory (vscode-file://vscode…task/Task.ts:922:10)
log.ts:460   ERR An unknown error occurred. Please consult the log for more details.
```

## Debug Information
Added debug logging shows that the file exists but contains an empty array:
```
[Roo-Debug] readApiMessages: Found API conversation history file, but it's empty. TaskId: 33369d03-10fb-436e-a708-ec3cdde71d80, Path: ~/.config/Code - Insiders/User/globalStorage/rooveterinaryinc.roo-cline/tasks/33369d03-10fb-436e-a708-ec3cdde71d80/api_conversation_history.json
```

## Reproduction Steps
1. Open a task
2. Open an orchestrator task from within that task
3. Close the orchestrator task
4. Attempt to resume the original task

## Potential Fix
PR #3772 (https://github.com/RooCodeInc/Roo-Code/pull/3772) may address this issue, though the exact cause is still unknown.

## Impact
This causes task resumption to fail with the error "Unexpected: No existing API conversation history" and prevents users from continuing their work.

--- a/src/core/task-persistence/apiMessages.ts
+++ b/src/core/task-persistence/apiMessages.ts
@@ -22,26 +22,41 @@ export async function readApiMessages({
 
 	if (await fileExistsAtPath(filePath)) {
 		const fileContent = await fs.readFile(filePath, "utf8")
-		const parsedData = JSON.parse(fileContent)
-		if (Array.isArray(parsedData) && parsedData.length === 0) {
+		try {
+			const parsedData = JSON.parse(fileContent)
+			if (Array.isArray(parsedData) && parsedData.length === 0) {
+				console.error(
+					`[Roo-Debug] readApiMessages: Found API conversation history file, but it's empty (parsed as []). TaskId: ${taskId}, Path: ${filePath}`,
+				)
+			}
+			return parsedData
+		} catch (error) {
 			console.error(
-				`[Roo-Debug] readApiMessages: Found API conversation history file, but it's empty. TaskId: ${taskId}, Path: ${filePath}`,
+				`[Roo-Debug] readApiMessages: Error parsing API conversation history file. TaskId: ${taskId}, Path: ${filePath}, Error: ${error}`,
 			)
+			throw error
 		}
-		return parsedData
 	} else {
 		const oldPath = path.join(taskDir, "claude_messages.json")
 
 		if (await fileExistsAtPath(oldPath)) {
 			const fileContent = await fs.readFile(oldPath, "utf8")
-			const parsedData = JSON.parse(fileContent)
-			if (Array.isArray(parsedData) && parsedData.length === 0) {
+			try {
+				const parsedData = JSON.parse(fileContent)
+				if (Array.isArray(parsedData) && parsedData.length === 0) {
+					console.error(
+						`[Roo-Debug] readApiMessages: Found OLD API conversation history file (claude_messages.json), but it's empty (parsed as []). TaskId: ${taskId}, Path: ${oldPath}`,
+					)
+				}
+				await fs.unlink(oldPath)
+				return parsedData
+			} catch (error) {
 				console.error(
-					`[Roo-Debug] readApiMessages: Found OLD API conversation history file (claude_messages.json), but it's empty. TaskId: ${taskId}, Path: ${oldPath}`,
+					`[Roo-Debug] readApiMessages: Error parsing OLD API conversation history file (claude_messages.json). TaskId: ${taskId}, Path: ${oldPath}, Error: ${error}`,
 				)
+				// DO NOT unlink oldPath if parsing failed, throw error instead.
+				throw error
 			}
-			await fs.unlink(oldPath)
-			return parsedData
 		}
 	}
 

--- a/src/core/task-persistence/apiMessages.ts
+++ b/src/core/task-persistence/apiMessages.ts
@@ -21,17 +21,34 @@ export async function readApiMessages({
 	const filePath = path.join(taskDir, GlobalFileNames.apiConversationHistory)
 
 	if (await fileExistsAtPath(filePath)) {
-		return JSON.parse(await fs.readFile(filePath, "utf8"))
+		const fileContent = await fs.readFile(filePath, "utf8")
+		const parsedData = JSON.parse(fileContent)
+		if (Array.isArray(parsedData) && parsedData.length === 0) {
+			console.error(
+				`[Roo-Debug] readApiMessages: Found API conversation history file, but it's empty. TaskId: ${taskId}, Path: ${filePath}`,
+			)
+		}
+		return parsedData
 	} else {
 		const oldPath = path.join(taskDir, "claude_messages.json")
 
 		if (await fileExistsAtPath(oldPath)) {
-			const data = JSON.parse(await fs.readFile(oldPath, "utf8"))
+			const fileContent = await fs.readFile(oldPath, "utf8")
+			const parsedData = JSON.parse(fileContent)
+			if (Array.isArray(parsedData) && parsedData.length === 0) {
+				console.error(
+					`[Roo-Debug] readApiMessages: Found OLD API conversation history file (claude_messages.json), but it's empty. TaskId: ${taskId}, Path: ${oldPath}`,
+				)
+			}
 			await fs.unlink(oldPath)
-			return data
+			return parsedData
 		}
 	}
 
+	// If we reach here, neither the new nor the old history file was found.
+	console.error(
+		`[Roo-Debug] readApiMessages: API conversation history file not found for taskId: ${taskId}. Expected at: ${filePath}`,
+	)
 	return []
 }
 

